### PR TITLE
feature: support pod use static vip

### PR DESCRIPTION
**Pull request by bobz965**
feature: support pod use static vip

[support-pod-use-static-vip.md](https://github.com/kubeovn/kube-ovn/files/8992165/support-pod-use-static-vip.md)

### CODE REVIEW

File: pkg/controller/pod.go [hunk 1]
```diff
@@ -1191,6 +1197,22 @@ func (c *Controller) acquireAddress(pod *v1.Pod, podNet *kubeovnNet) (string, st
...
+	isStsPod, _ := isStatefulSetPod(pod)
+	// if pod has static vip
+	vipName := pod.Annotations[util.VipAnnotation]
+	if vipName != "" {
+		vip, err := c.virtualIpsLister.Get(vipName)
+		if err != nil {
+			klog.Errorf("failed to get static vip '%s', %v", vipName, err)
+			return "", "", "", podNet.Subnet, err
+		}
+		nicName := ovs.PodNameToPortName(podName, pod.Namespace, podNet.ProviderName)
```
Comment (oilbeater) [inline pkg/controller/pod.go:1209]: nicName -> portName

File: pkg/controller/vip.go [hunk 2]
```diff
@@ -508,3 +512,79 @@ func (c *Controller) patchVipStatus(key, v4ip string, ready bool) error {
 	}
 	return nil
 }
+
+func (c *Controller) podReuseVip(key, podName string, isStsPod bool) error {
```
Comment (oilbeater) [inline pkg/controller/vip.go:516]: podName -> portName

File: pkg/controller/vip.go [hunk 3]
```diff
@@ -508,3 +512,79 @@ func (c *Controller) patchVipStatus(key, v4ip string, ready bool) error {
...
+	var op string
+
+	if vip.Labels[util.IpReservedLabel] != "" {
+		if isStsPod && vip.Labels[util.IpReservedLabel] == podName {
+			return nil
+		} else {
+			return fmt.Errorf("vip '%s' is in use by pod %s", vip.Name, vip.Labels[util.IpReservedLabel])
+		}
+	}
+	if podName != "" {
```
Comment (oilbeater) [inline pkg/controller/vip.go:537]: when the podName is empty?

File: pkg/controller/vip.go [hunk 4]
```diff
@@ -508,3 +512,79 @@ func (c *Controller) patchVipStatus(key, v4ip string, ready bool) error {
...
+			klog.Errorf("failed to patch label for vip '%s', %v", vip.Name, err)
+			return err
+		}
+		c.ipam.ReleaseAddressByPod(key)
+		c.updateSubnetStatusQueue.Add(vip.Spec.Subnet)
+	}
+	return nil
+}
+
+func (c *Controller) podNotReuseVip(key string) error {
```
Comment (oilbeater) [inline pkg/controller/vip.go:557]: podNotReuseVip  -> releaseVip

**Review by oilbeater — CHANGES_REQUESTED** [review]

**Comment by oilbeater** [comment]
We should check if the allocated vip still in use at gc.

File: pkg/controller/pod.go (hunk 1 above)
Comment (bobz965) [reply pkg/controller/pod.go]: ok

File: pkg/controller/vip.go (hunk 3 above)
Comment (bobz965) [reply pkg/controller/vip.go]: remove this if

File: pkg/controller/vip.go (hunk 4 above)
Comment (bobz965) [reply pkg/controller/vip.go]: ok

**Comment by bobz965** [comment]
> We should check if the allocated vip is still in use at gc.

add vip gc

**Review by oilbeater — APPROVED** [review]

**Pull request merged**